src/instruction_graph_generator.cc

```diff
@@ -0,0 +1,2038 @@
+	}
+
+	// Receive all remote data (which overlaps with the accessed region) into host memory
+	std::vector<region<3>> all_concurrent_reads;
+	for(const auto& reads : concurrent_reads_from_memory) {
+		all_concurrent_reads.insert(all_concurrent_reads.end(), reads.begin(), reads.end());
+	}
+	for(const auto& receive : applied_receives) {
+		commit_pending_region_receive_to_host_memory(current_batch, bid, receive, all_concurrent_reads);
+	}
+
+	// Create the necessary coherence copy instructions to satisfy all remaining requirements locally. The iterations of this loop are independent with the
+	// notable exception of host_memory_id in the presence of staging copies: There we rely on the fact that `host_memory_id < device_memory_id` to allow
+	// coherence copies to device memory to create device -> host -> device copy chains.
+	for(memory_id mid = 0; mid < concurrent_reads_from_memory.size(); ++mid) {
+		establish_coherence_between_buffer_memories(current_batch, bid, mid, concurrent_reads_from_memory[mid]);
+	}
+}
+
+local_reduction generator_impl::prepare_task_local_reduction(
+    batch& command_batch, const reduction_info& rinfo, const execution_command& ecmd, const task& tsk, const size_t num_concurrent_chunks) //
+{
+	const auto [rid_, bid_, reduction_task_includes_buffer_value] = rinfo;
+	const auto bid = bid_; // allow capturing in lambda
+
+	auto& buffer = m_buffers.at(bid);
+
+	local_reduction red;
+	red.include_local_buffer_value = reduction_task_includes_buffer_value && ecmd.is_reduction_initializer();
+	red.current_value_offset = red.include_local_buffer_value ? 1 : 0;
+	red.num_input_chunks = red.current_value_offset + num_concurrent_chunks;
+	red.chunk_size_bytes = scalar_reduction_box.get_area() * buffer.elem_size;
+
+	// If the reduction only has a single local contribution, we simply accept it as the fully-reduced final value without issuing additional instructions.
+	// A local_reduction with num_input_chunks == 1 is treated as a no-op by `finish_task_local_reduction`.
+	assert(red.num_input_chunks > 0);
+	if(red.num_input_chunks == 1) return red;
+
+	// Create a gather-allocation into which `finish_task_local_reduction` will copy each new partial result, and if the reduction is not
+	// initialize_to_identity, we copy the current buffer value before it is being overwritten by the kernel.
+	red.gather_aid = new_allocation_id(host_memory_id);
+	red.gather_alloc_instr = create<alloc_instruction>(
+	    command_batch, red.gather_aid, red.num_input_chunks * red.chunk_size_bytes, buffer.elem_align, [&](const auto& record_debug_info) {
+		    record_debug_info(
+		        alloc_instruction_record::alloc_origin::gather, buffer_allocation_record{bid, buffer.debug_name, scalar_reduction_box}, red.num_input_chunks);
+	    });
+	add_dependency(red.gather_alloc_instr, m_last_epoch, instruction_dependency_origin::last_epoch);
+
+	if(red.include_local_buffer_value) {
```

Comment:
I think the source of confusion is the difference between a _kernel chunk_ and a _reduction chunk_. If the local node is the single reduction initializer in the cluster, it gains the initial value as an additional reduction chunk (which is gather-copied before the device-side reduction starts), so in case of a single kernel chunk there will be 2 reduction chunks and a local reduction instruction is emitted.

Reply:
Hmm ok, yes - not sure how I feel about the term "chunk" being overloaded here, but it makes more sense now!